v2/pkg/engine/jsonschema/schema.go

```diff
@@ -0,0 +1,250 @@
+package jsonschema
+
+import (
+	"encoding/json"
+)
+
+// SchemaType represents the type of a JSON Schema property
+type SchemaType string
+
+const (
+	TypeObject  SchemaType = "object"
+	TypeArray   SchemaType = "array"
+	TypeString  SchemaType = "string"
+	TypeNumber  SchemaType = "number"
+	TypeInteger SchemaType = "integer"
+	TypeBoolean SchemaType = "boolean"
+	TypeNull    SchemaType = "null"
+)
+
+// JsonSchema represents a JSON Schema definition
+type JsonSchema struct {
+	// Core schema fields
+	Type                 SchemaType             `json:"type,omitempty"`
+	Properties           map[string]*JsonSchema `json:"properties,omitempty"`
+	Required             []string               `json:"required,omitempty"`
+	AdditionalProperties *bool                  `json:"additionalProperties,omitempty"`
+	Description          string                 `json:"description,omitempty"`
+	Nullable             bool                   `json:"nullable,omitempty"`
+
+	// Array-specific fields
+	Items *JsonSchema `json:"items,omitempty"`
+
+	// Enum values
+	Enum []interface{} `json:"enum,omitempty"`
+
+	// Default value
+	Default interface{} `json:"default,omitempty"`
+
+	// String-specific fields
+	Format string `json:"format,omitempty"`
+
+	// Number-specific fields
+	Minimum *float64 `json:"minimum,omitempty"`
+	Maximum *float64 `json:"maximum,omitempty"`
+
+	// Additional validation
+	Pattern string `json:"pattern,omitempty"`
+}
+
+// MarshalJSON customizes JSON serialization to omit empty fields
+func (s *JsonSchema) MarshalJSON() ([]byte, error) {
```

Comment:
You could also achieve this by making things a pointer and then have "omitempty"

Reply:
This leads to a bad api. I'm going to keep it